modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java

```diff
@@ -128,11 +115,30 @@ private void waitForReindexDone(String taskId, ActionListener<StartReindexJobAct
                 @Override
                 public void onResponse(PersistentTasksCustomMetaData.PersistentTask<ReindexJob> task) {
                     ReindexJobState state = (ReindexJobState) task.getState();
-                    if (state.getJobException() == null) {
-                        listener.onResponse(new StartReindexJobAction.Response(taskId, state.getReindexResponse()));
+                    if (state.getStatus() == ReindexJobState.Status.FAILED_TO_READ_FROM_REINDEX_INDEX) {
+                        listener.onFailure(new ElasticsearchException("Reindexing failed. Task node could not read from "
+                            + ReindexIndexClient.REINDEX_INDEX + " index"));
+                    } else if (state.getStatus() == ReindexJobState.Status.FAILED_TO_WRITE_TO_REINDEX_INDEX) {
+                        listener.onFailure(new ElasticsearchException("Reindexing failed. Task node could not write result to "
+                            + ReindexIndexClient.REINDEX_INDEX + " index"));
                     } else {
-                        listener.onFailure(state.getJobException());
+                        reindexIndexClient.getReindexTaskDoc(taskId, new ActionListener<>() {
+                            @Override
+                            public void onResponse(ReindexTaskIndexState reindexState) {
+                                if (reindexState.getException() == null) {
+                                    listener.onResponse(new StartReindexJobAction.Response(taskId, reindexState.getReindexResponse()));
+                                } else {
+                                    listener.onFailure(reindexState.getException());
+                                }
+                            }
```

Comment:
Should we then delete the persistent task, given that we responded and everything is done? Probably to be done in a follow-up rather than in this PR.